File: g25/spec_conf_csharp/ReportUsage.cs
/*
Copyright (C) 2008 Some Random Person
*/
/*
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

*/
using System;
using System.Collections.Generic;
using System.Text;
namespace c3ga_ns {


class ReportUsage
{
    protected static Dictionary<string, ReportUsage> s_reportUsage = new Dictionary<string, ReportUsage>();

    /// <summary>
    ///  Adds a report to the global report usage hash table.
    /// </summary>
    public static void MergeReport(ReportUsage RU)
    {
        if (s_reportUsage.ContainsKey(RU.GetReportString())) {
            s_reportUsage[RU.GetReportString()].IncrementCount();
        }
        else {
             s_reportUsage.Add(RU.GetReportString(), RU);
        }
    }

    /// prints out all reports, sorted by m_str
    public static string GetReport(bool includeCount)
    {
        StringBuilder sb = new StringBuilder();
        SortedDictionary<string, ReportUsage> sortedReportUsage = new SortedDictionary<string, ReportUsage>(s_reportUsage);
        foreach (KeyValuePair<string, ReportUsage> kvp in sortedReportUsage ) {
            ReportUsage ru = kvp.Value;
            sb.Append(ru.GetReportString());
            if (includeCount)
            {
                sb.AppendFormat("  <!-- used %d time%s -->", ru.GetReportCount(), (ru.GetReportCount() == 1) ? "" : "s");
            }
            sb.AppendLine();
        }

        return sb.ToString();
    }

    public ReportUsage(string xmlStr)
    {
        m_str = xmlStr;
        m_count = 1;
    }

    public string GetReportString()
    {
        return m_str;
    }

    public int GetReportCount()
    {
        return m_count;
    }

    protected void IncrementCount()
    {
        m_count++;
    }

    /// <summary>
    /// XML string describing the report.
    /// </summary>
    protected string m_str;

    /// <summary>
    /// Number of times report has been reported.
    /// </summary>
    int m_count;
}

} // end of namespace c3ga_ns

File: usage_example/java/Step3/src/C3ga_pkg/ReportUsage.java
/*
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

*/
package C3ga_pkg;
import java.util.Hashtable;
import java.util.TreeSet;
import java.util.Map;


public class ReportUsage implements Comparable<ReportUsage>
{
    protected static Hashtable<String, ReportUsage> s_reportUsage = new Hashtable<String, ReportUsage>();

    /// <summary>
    ///  Adds a report to the global report usage hash table.
    /// </summary>
    public static void mergeReport(ReportUsage RU)
    {
		// Note: s_reportUsage (Hashtable<>) is thread-safe, so no locking required.
        if (s_reportUsage.containsKey(RU.getReportString())) {
            s_reportUsage.get(RU.getReportString()).incrementCount();
        }
        else {
             s_reportUsage.put(RU.getReportString(), RU);
        }
    }

    /// prints out all reports, sorted by m_str
    public static String getReport(boolean includeCount)
    {
        StringBuilder sb = new StringBuilder();
        
        TreeSet<ReportUsage> sortedSet = new TreeSet<ReportUsage>();
        for (Map.Entry<String, ReportUsage> kvp : s_reportUsage.entrySet()) {
        	sortedSet.add(kvp.getValue());
        }
        
        for (ReportUsage ru : sortedSet) {
            sb.append(ru.getReportString());
            if (includeCount)
            {
                sb.append("  <!-- used " + ru.getReportCount() + " time" + ((ru.getReportCount() == 1) ? "" : "s") + " -->");
            }
            sb.append("\n");
        }
        
        if (s_reportUsage.isEmpty()) {
				sb.append("  <!-- no general multivector function usage -->");
		}

        return sb.toString();
    }

    public ReportUsage(String xmlStr)
    {
        m_str = xmlStr;
        m_count = 1;
    }

	@Override
	public int compareTo(ReportUsage arg0) {
		
		return m_str.compareTo(((ReportUsage)arg0).getReportString());
	}
	
    public String getReportString()
    {
        return m_str;
    }

    public int getReportCount()
    {
        return m_count;
    }

    protected void incrementCount()
    {
        m_count++;
    }

    /// <summary>
    /// XML string describing the report.
    /// </summary>
    protected String m_str;

    /// <summary>
    /// Number of times report has been reported.
    /// </summary>
    protected int m_count;
}

